### PR TITLE
Make PKCE implementation compliant with RFC7636

### LINKED ISSUE #316: Enforce compliant PKCE code_verifier lengths
https://datatracker.ietf.org/doc/html/rfc7636#section-4.1 says that valid code_verifier is 43 to 128 characters. Currently the length isn't validated.

The expected response for invalid length is:

`{"error":"invalid_grant","error_description":"PKCE invalid code verifier"}`

### PR DESCRIPTION
**Pull request by sandhose**
This checks for the PKCE code_verifier length as well as the characters
used. It also give better errors when the PKCE verifier is invalid.

Fixes #316

### CODE REVIEW

File: crates/data-model/src/oauth2/authorization_grant.rs [hunk 1]
```diff
@@ -24,6 +24,21 @@ use url::Url;
 use super::{client::Client, session::Session};
 use crate::{traits::StorageBackend, StorageBackendMarker};
 
+#[derive(Debug, Error)]
+pub enum PkceVerificationError {
+    #[error("code_verifier should be at least 43 characters long")]
```
Comment (sandhose) [inline crates/data-model/src/oauth2/authorization_grant.rs:29]: So here the `error_description` would look like that:

```
PKCE verification failed: code_verifier should be at least 43 characters long
```

File: crates/data-model/src/oauth2/authorization_grant.rs [hunk 2]
```diff
@@ -39,9 +54,27 @@ impl Pkce {
...
 
-    #[must_use]
-    pub fn verify(&self, verifier: &str) -> bool {
-        self.challenge_method.verify(&self.challenge, verifier)
+    pub fn verify(&self, verifier: &str) -> Result<(), PkceVerificationError> {
+        if verifier.len() < 43 {
+            return Err(PkceVerificationError::TooShort);
+        }
+
+        if verifier.len() > 43 {
```
Comment (hughns) [inline crates/data-model/src/oauth2/authorization_grant.rs:62]:
```suggestion
        if verifier.len() > 128 {
```

**Review by hughns — CHANGES_REQUESTED** [review]
Condition for max length is wrong. Otherwise looks good.

**Comment by sandhose** [comment]
Actually the allowed characters check was also buggy. Fixed it and added a bunch of unit tests

**Review by hughns — APPROVED** [review]

**Pull request merged**